### PR TITLE
Change `expect` messages to `Result` in `forc doc`

### LINKED ISSUE #3655: Convert `expect()` messages in `forc doc` to `anyhow::Result`
import `anyhow::Result` where necessary and remove the `expect` messages in favor of `Result`

### PR DESCRIPTION
**Pull request by gr00vytvniks**
Work in progress, implementing `anyhow::Result` and handling `expect` & `unwrap` where necessary

Closes #3655

### CODE REVIEW

**Review by bingcicle — CHANGES_REQUESTED** [review]
Thanks for the PR! :) Just a few nits

File: forc-plugins/forc-doc/src/descriptor.rs
Comment (kayagokalp) [inline forc-plugins/forc-doc/src/descriptor.rs:95]: ^^ Ditto

File: forc-plugins/forc-doc/src/descriptor.rs
Comment (kayagokalp) [inline forc-plugins/forc-doc/src/descriptor.rs:132]: ^^ Ditto

File: forc-plugins/forc-doc/src/descriptor.rs
Comment (kayagokalp) [inline forc-plugins/forc-doc/src/descriptor.rs:169]: ^^ Ditto

File: forc-plugins/forc-doc/src/descriptor.rs
Comment (kayagokalp) [inline forc-plugins/forc-doc/src/descriptor.rs:205]: ^^ Ditto

File: forc-plugins/forc-doc/src/descriptor.rs
Comment (kayagokalp) [inline forc-plugins/forc-doc/src/descriptor.rs:267]: ^^ Ditto

**Review by kayagokalp — CHANGES_REQUESTED** [review]
Thanks for the contribution👌, I left some comments, suggestions to clean up some of the things a little, also agree with @bingcicle's comments in the review. 

As a discussion point:

I generally tend to use `anyhow::Result` where the source of error is user related and `expect` in places where I expect something to be there and if it is not it is due to an error in the code logic. So some of the things here might be better of left as `expect` imo but I am good with it either way 👍 I am not aware if this is the idiomatic way of deciding between `expect` and `Result` more like a rule I tend to follow without being explicit about it

**Comment by gr00vytvniks** [comment]
@kayagokalp @bingcicle  It doesn't look like I can request both of your reviews at the same time so i'll just tag you both here

**Review by bingcicle — CHANGES_REQUESTED** [review]

File: forc-plugins/forc-doc/src/doc.rs
Comment (kayagokalp) [reply forc-plugins/forc-doc/src/doc.rs:141]: Yepp, seems like I missed it sorry for this one @gr00vytvniks, let's go back to `match` and +1 on the otter comment too

**Comment by eureka-cpu** [comment]
Found a bug that is unrelated to this #3923

**Comment by eureka-cpu** [comment]
We should probably get this in sooner rather than later lest we want more merge conflicts. PR is up for the bug mentioned above: #3930

File: forc-plugins/forc-doc/src/doc.rs
Comment (bingcicle) [inline forc-plugins/forc-doc/src/doc.rs:184]: Sorry if I wasn't clear with what I meant by "going back to our original matching", but this could be what you originally went with in your initial PR:

```rust
match self.0.last() {
  Some(location) => Ok(location),
  None => bail!("There will always be at least the module location"),
```

Note also the err message change, as pointed out [here](https://github.com/FuelLabs/sway/pull/3691#discussion_r1073313586).

File: forc-plugins/forc-doc/src/doc.rs
Comment (bingcicle) [inline forc-plugins/forc-doc/src/doc.rs:192]: Same as above:

```suggestion
        match self.0.first() {
          Some(name) => Ok(name),
          None => bail!("Project name missing"),
        }
```

**Review by bingcicle — CHANGES_REQUESTED** [review]
Thank you for your patience @gr00vytvniks ! Just last few things here.

File: forc-plugins/forc-doc/src/doc.rs
Comment (eureka-cpu) [reply forc-plugins/forc-doc/src/doc.rs:184]: Hmm I think for now the expect message here is okay. Implementing result there means you'd have to worry about how you handle every call to location in the render macro which doesn't return a result. There's a few for loops in there where I can see it would be difficult to find a non-hacky way of resolving the results before using them in the `box_html!` macro.

Comment (bingcicle) [reply forc-plugins/forc-doc/src/doc.rs:184]: that's reasonable!

**Review by bingcicle — DISMISSED** [review]
Thank you @gr00vytvniks, great work! Giving this the :+1:

**Review by eureka-cpu — APPROVED** [review]
🚢

**Review by bingcicle — APPROVED** [review]
:shipit:

**Comment by mohammadfawaz** [comment]
@kayagokalp this is waiting on a re-review from you :)

**Comment by kayagokalp** [comment]
Ahh must have missed this one! thanks for pinging @mohammadfawaz 👍

**Review by kayagokalp — APPROVED** [review]

**Pull request merged**
